src/Microsoft.Health.Dicom.Core/Features/Export/IdentifierExportSource.cs

```diff
@@ -0,0 +1,120 @@
+        _options = EnsureArg.IsNotNull(options?.Value, nameof(options));
+    }
+
+    public ValueTask DisposeAsync()
+        => ValueTask.CompletedTask;
+
+    public async IAsyncEnumerator<ReadResult> GetAsyncEnumerator(CancellationToken cancellationToken = default)
+    {
+        IEnumerable<ReadResult> results = Enumerable.Empty<ReadResult>();
+        for (int i = _startIndex; i < _options.Values.Count; i++)
+        {
+            // Identifier has already been validated
+            var identifier = DicomIdentifier.Parse(_options.Values[i]);
+
+            // Attempt to read the data
+            IReadOnlyList<VersionedInstanceIdentifier> instances = identifier.Type switch
+            {
+                ResourceType.Study => await _instanceStore.GetInstanceIdentifiersInStudyAsync(_partition.PartitionKey, identifier.StudyInstanceUid, cancellationToken),
+                ResourceType.Series => await _instanceStore.GetInstanceIdentifiersInSeriesAsync(_partition.PartitionKey, identifier.StudyInstanceUid, identifier.SeriesInstanceUid, cancellationToken),
+                _ => await _instanceStore.GetInstanceIdentifierAsync(_partition.PartitionKey, identifier.StudyInstanceUid, identifier.SeriesInstanceUid, identifier.SopInstanceUid, cancellationToken),
+            };
+
+            if (instances.Count > 0)
+            {
+                foreach (VersionedInstanceIdentifier read in instances)
+                    yield return ReadResult.ForIdentifier(read);
+            }
+            else
+            {
+                var args = new ReadFailureEventArgs(
+                    identifier,
+                    new FileNotFoundException(
+                        identifier.Type switch
+                        {
+                            ResourceType.Study => DicomCoreResource.StudyNotFound,
+                            ResourceType.Series => DicomCoreResource.SeriesNotFound,
+                            _ => DicomCoreResource.InstanceNotFound,
+                        }));
+
+                ReadFailure?.Invoke(this, args);
+                yield return ReadResult.ForFailure(args);
+            }
+        }
+    }
+
+    public bool TryDequeueBatch(int size, out TypedConfiguration<ExportSourceType> batch)
+    {
+        EnsureArg.IsGt(size, 0, nameof(size));
+
+        batch = GetConfiguration(size, remove: true);
+        return batch != null;
+    }
+
+    private TypedConfiguration<ExportSourceType> GetConfiguration(int? maxSize = null, bool remove = false)
+    {
+        int count = maxSize.HasValue
+            ? Math.Min(maxSize.GetValueOrDefault(), _options.Values.Count - _startIndex)
+            : _options.Values.Count - _startIndex;
+
+        if (count == 0)
+            return null;
+
+        // Create a configuration that describes the source over this subset
+        var source = new TypedConfiguration<ExportSourceType>
+        {
+            Configuration = new ConfigurationBuilder().AddInMemoryCollection().Build(),
```

Comment:
For my understanding, What is this doing? Feels too much for every batch get.

Reply:
Yeah, it might be a bit much. If I recall the type names correctly, I think it's creating a `InMemoryConfigurationProvider` which has an `InMemoryConfigurationSource` which ultimately uses a `Dictionary<string, string> for storing the data lol.

I think I can avoid some of this overhead by changing the API